### PR TITLE
Implement package syncing

### PR DESCRIPTION
**Pull request by tigrannajaryan**
- This implements "Packages" spec section https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#packages-1
- TODO: increase test coverage to handle many more edge cases.
- TODO: add package usage to the example Agent/Server.

### CODE REVIEW

File: client/internal/packagessyncer.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+			if err != nil {
+				lastErr = err
+			}
+		}
+	}
+
+	// Also remove packages that were not offered from the statuses.
+	for name := range s.statuses.Packages {
+		if _, offered := s.available.Packages[name]; !offered {
+			delete(s.available.Packages, name)
```
Comment (andykellr) [inline client/internal/packagessyncer.go:299]:
```suggestion
			delete(s.statuses.Packages, name)
```

**Comment by andykellr** [comment]
Aside from the small issue I mentioned, overall this looks good. However, unless I missed something, it appears to me that it is possible to have multiple sync operations running at the same time which could be confusing.

**Comment by tigrannajaryan** [comment]
>However, unless I missed something, it appears to me that it is possible to have multiple sync operations running at the same time which could be confusing.

Good point. I will create an issue to fix this in a future PR.

File: client/internal/packagessyncer.go (hunk 1 above)
Comment (tigrannajaryan) [reply client/internal/packagessyncer.go]: @andykellr fixed. PTAL.

File: client/clientimpl_test.go [hunk 2]
```diff
@@ -966,3 +968,253 @@ func TestRemoteConfigUpdate(t *testing.T) {
...
+		srv := internal.StartMockServer(t)
+		srv.EnableExpectMode()
+
+		localPackageState := internal.NewInMemPackagesStore()
+
+		var syncerDoneCh <-chan struct{}
+
+		// Prepare a callback that returns either success or failure.
+		onPackagesAvailable := func(ctx context.Context, packages *protobufs.PackagesAvailable, syncer types.PackagesSyncer) error {
+			if testCase.errorOnCallback {
```
Comment (pmm-sumo) [inline client/clientimpl_test.go:995]: Why the error is returned when the flag is false rather than true?

Comment (tigrannajaryan) [reply client/clientimpl_test.go:995]: Oops. Fixed.

File: client/internal/httpsender.go [hunk 3]
```diff
@@ -97,6 +103,10 @@ func (h *HTTPSender) makeOneRequestRoundtrip(ctx context.Context) {
 	if err != nil {
 		return
```
Comment (pmm-sumo) [inline client/internal/httpsender.go:104]: not scope of this PR but just noticed this - shouldn't we at least log the error?

**Review by pmm-sumo — APPROVED** [review]

**Comment by kpratyus** [comment]
In the file packagesyncer.go defer func() closes the channel s.doneCh but it doesn't necessarily wait for the subroutine go 
s.doSync to get completed so wouldn't it be giving a wrong sync status?

```
func (s *packagesSyncer) Sync(ctx context.Context) error {	
         defer func() {
		close(s.doneCh)
	}()'

	// Prepare package statuses.
	if err := s.initStatuses(); err != nil {
		return err
	}

	if err := s.clientSyncedState.SetPackageStatuses(s.statuses); err != nil {
		return err
	}

	// Now do the actual syncing in the background.
	go s.doSync(ctx, s.packageSyncComplete)

	return nil
}
```
